spccore/internal/pool_provider.py

```diff
@@ -0,0 +1,90 @@
+"""
+This module allow us to simplify the logic of any method that can run with either multiple threads or single thread.
+
+To use this wrapper with multiple threads::
+    pool = pool_provider.get_pool()
+    try:
+        pool.map(function, iterable)
+    finally:
+        pool.terminate()
+
+To use this wrapper for single thread::
+    pool = pool_provider.get_pool(size=1)
+
+"""
+
+import multiprocessing
+import multiprocessing.dummy
+
+
+DEFAULT_POOL_SIZE = 8
+_single_threaded = False
+
+
+class SingleThreadPool:
```

Comment:
make 2 immutable pool providers that are either capable of only providing multi-thread or providing single thread.